semimpl/do_test.go

```diff
@@ -0,0 +1,177 @@
+package semimpl_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/benchttp/runner/semimpl"
+)
+
+func TestDo(t *testing.T) {
+	t.Run("stop until maxIter is reached", func(t *testing.T) {
+		var (
+			numWorkers = 1
+			gotIter    = 0
+			maxIter    = 10
+			expIter    = 10
+		)
+
+		semimpl.Do(context.Background(), numWorkers, maxIter, func() {
+			gotIter++
+		})
+
+		if gotIter != expIter {
+			t.Errorf("iterations: exp %d, got %d", expIter, gotIter)
+		}
+	})
+
+	t.Run("stop on context timeout", func(t *testing.T) {
+		var (
+			timeout    = 50 * time.Millisecond
+			interval   = timeout / 5
+			numWorkers = 1
+			maxIter    = 0 // infinite iterations
+
+			margin      = 15 * time.Millisecond
+			maxDuration = timeout + margin
+		)
+
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
+
+		gotDuration := timeFunc(func() {
+			semimpl.Do(ctx, numWorkers, maxIter, func() {
+				time.Sleep(interval)
+			})
+		})
+
+		if gotDuration > maxDuration {
+			t.Errorf(
+				"context timeout duration: exp < %dms, got %dms",
+				maxDuration.Milliseconds(), gotDuration.Milliseconds(),
+			)
+		}
+	})
+
+	t.Run("stop on context cancel", func(t *testing.T) {
+		var (
+			timeout    = 50 * time.Millisecond
+			interval   = 10 * time.Millisecond
+			numWorkers = 1
+			maxIter    = 0 // infinite iterations
```

Comment:
[suggestion]
Is there a reason you prefer to use `0` for maxIter instead of `nil`?
It makes a little more sense to me that `nil` would mean that there is no maximum, when 0 seems strange (but it may be because I'm not used to Go).

Reply:
Yup, `int` can't be `nil`. Only a few non-primitives types can: maps, slices (not arrays), funcs, interfaces, pointers (I'm probably missing some). They have instead a default value: `0` for numbers, `""` for strings, `false` for bools...
Not having to deal with `nil` is actually quite handy, but the downside is you can't know whether a value has been initialized or defaulted.

Reply:
Ah, I forgot this... Ok, 0 it is then 😄